src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala

```diff
@@ -0,0 +1,439 @@
+import org.broadinstitute.dsde.rawls.model.WorkspaceJsonSupport._
+import org.broadinstitute.dsde.rawls.model._
+import org.broadinstitute.dsde.rawls.openam.MockOpenAmDirectives
+import org.broadinstitute.dsde.rawls.workspace.WorkspaceService
+import org.scalatest.{FlatSpec, Matchers}
+import spray.http.HttpHeaders.Cookie
+import spray.http._
+import spray.httpx.SprayJsonSupport._
+import spray.json._
+import spray.routing.HttpService
+import spray.testkit.ScalatestRouteTest
+
+import scala.concurrent.duration._
+
+/**
+ * Created by dvoet on 4/24/15.
+ */
+class WorkspaceApiServiceSpec extends FlatSpec with HttpService with ScalatestRouteTest with Matchers with OrientDbTestFixture {
+  // increate the timeout for ScalatestRouteTest from the default of 1 second, otherwise
+  // intermittent failures occur on requests not completing in time
+  implicit val routeTestTimeout = RouteTestTimeout(5.seconds)
+
+  // these tokens won't work for login to remote services: that requires a password and is therefore limited to the integration test
+  def addOpenAmCookie(token: String) = addHeader(Cookie(HttpCookie("iPlanetDirectoryPro", token)))
+  def addMockOpenAmCookie = addOpenAmCookie("test_token")
+
+  def actorRefFactory = system
+
+  val mockServer = RemoteServicesMockServer()
+
+  override def beforeAll() = {
+    super.beforeAll
+    mockServer.startServer
+  }
+
+  override def afterAll() = {
+    super.afterAll
+    mockServer.stopServer
+  }
+
+  case class TestApiService(dataSource: DataSource) extends WorkspaceApiService with EntityApiService with MethodConfigApiService with SubmissionApiService with GoogleAuthApiService with MockOpenAmDirectives {
+    def actorRefFactory = system
+
+    val submissionSupervisor = system.actorOf(SubmissionSupervisor.props(
+      new GraphWorkspaceDAO(),
+      new GraphSubmissionDAO(new GraphWorkflowDAO()),
+      new HttpExecutionServiceDAO(mockServer.mockServerBaseUrl),
+      new GraphWorkflowDAO(),
+      new GraphEntityDAO(),
+      new GraphMethodConfigurationDAO(),
+      dataSource
+    ).withDispatcher("submission-monitor-dispatcher"), "test-wsapi-submission-supervisor")
+    val workspaceServiceConstructor = WorkspaceService.constructor(dataSource, workspaceDAO, entityDAO, methodConfigDAO, new HttpMethodRepoDAO(mockServer.mockServerBaseUrl), new HttpExecutionServiceDAO(mockServer.mockServerBaseUrl), MockGoogleCloudStorageDAO, submissionSupervisor, submissionDAO)_
+
+    def cleanupSupervisor = {
+      submissionSupervisor ! PoisonPill
+    }
+  }
+
+  def withApiServices(dataSource: DataSource)(testCode: TestApiService => Any): Unit = {
+    val apiService = new TestApiService(dataSource)
+    try {
+      testCode(apiService)
+    } finally {
+      apiService.cleanupSupervisor
+    }
+  }
+
+  def withTestDataApiServices(testCode: TestApiService => Any): Unit = {
+    withDefaultTestDatabase { dataSource =>
+      withApiServices(dataSource)(testCode)
+    }
+  }
```

Comment:
our service specs have an awful lot of similarity in sections like this that we can probably reuse:

```
def actorRefFactory = system

    val submissionSupervisor = system.actorOf(SubmissionSupervisor.props(
      new GraphWorkspaceDAO(),
      new GraphSubmissionDAO(new GraphWorkflowDAO()),
      new HttpExecutionServiceDAO(mockServer.mockServerBaseUrl),
      new GraphWorkflowDAO(),
      new GraphEntityDAO(),
      new GraphMethodConfigurationDAO(),
      dataSource
    ).withDispatcher("submission-monitor-dispatcher"), "test-wsapi-submission-supervisor")
    val workspaceServiceConstructor = WorkspaceService.constructor(dataSource, workspaceDAO, entityDAO, methodConfigDAO, new HttpMethodRepoDAO(mockServer.mockServerBaseUrl), new HttpExecutionServiceDAO(mockServer.mockServerBaseUrl), MockGoogleCloudStorageDAO, submissionSupervisor, submissionDAO)_

    def cleanupSupervisor = {
      submissionSupervisor ! PoisonPill
    }
  }

  def withApiServices(dataSource: DataSource)(testCode: TestApiService => Any): Unit = {
    val apiService = new TestApiService(dataSource)
    try {
      testCode(apiService)
    } finally {
      apiService.cleanupSupervisor
    }
  }

  def withTestDataApiServices(testCode: TestApiService => Any): Unit = {
    withDefaultTestDatabase { dataSource =>
      withApiServices(dataSource)(testCode)
    }
  }
```